cloudigrade/analyzer/tasks.py

```diff
@@ -0,0 +1,176 @@
+"""Celery tasks for analyzing incoming logs."""
+import json
+import logging
+from typing import Any, Dict
+
+from django.conf import settings
+from django.db import transaction
+from django.utils.translation import gettext as _
+
+from account.models import AwsAccount, InstanceEvent
+from account.util import save_instance_events, save_machine_images, \
+    start_image_inspection, tag_windows
+from util import aws
+from util.aws import is_instance_windows, rewrap_aws_errors
+from util.celery import retriable_shared_task
+
+logger = logging.getLogger(__name__)
+
+
+@retriable_shared_task
+@rewrap_aws_errors
+def analyze_log():
+    """Read SQS Queue for log location, and parse log for events."""
+    queue_url = settings.CLOUDTRAIL_EVENT_URL
+
+    logs = []
+    extracted_messages = []
+    instances = {}
+
+    # Get messages off of an SQS queue
+    messages = aws.receive_message_from_queue(queue_url)
+
+    # Parse the SQS messages to get S3 object locations
+    for message in messages:
+        extracted_messages.extend(aws.extract_sqs_message(message))
+
+    # Grab the object contents from S3
+    for extracted_message in extracted_messages:
+        bucket = extracted_message['bucket']['name']
+        key = extracted_message['object']['key']
+        logs.append(aws.get_object_content_from_s3(bucket, key))
+
+    # Parse logs for on/off events
+    for log in logs:
+        if log:
+            instances = _parse_log_for_ec2_events(log)
+
+    if instances:
+        _save_results(instances)
+        logger.debug(_('Saved instances and/or events to the DB.'))
+    else:
+        logger.debug(_('No instances or events to save to the DB.'))
+
+    aws.delete_message_from_queue(queue_url, messages)
+
+
+def _parse_log_for_ec2_events(log):
+    """
+    Parse S3 log for EC2 on/off events.
+
+    Args:
+        log (string): The string contents of the log file.
+
+    Returns:
+        list(tuple): List of instance data seen in log.
+        list(tuple): List of instance_event data seen in log.
+
+    """
+    ec2_event_map: Dict[str, Any] = {
```

Comment:
I wonder if this should be moved up in the module as a pseudo-constant since it doesn't need to be redefined on every function call.

Reply:
Also, interesting use of type hinting! I haven't checked, but I think this might be the first type hint in cloudigrade. What was your rationale for using it here?

Reply:
The type hinting was a left over from the code originally written by Andrew, upon a second look we don't need it, so it's gone 🎉